File: binary-search-trees/BST-class.js
// Average
    // time: O(log(n))
    // space: O(1)
// Worst
  // time: O(n)
  // space: O(1)
// where n is the number of nodes in the tree
class BST {
  constructor(value) {
    this.value = value
    this.left = null
    this.right = null
  }

  insert(value) {
		let current = this
		while (true) {
			if (value < current.value) {
				if (current.left === null) {
					current.left = new BST(value)
					break
				} else {
					current = current.left
				}
			} else {
				if (current.right === null) {
					current.right = new BST(value)
					break
				} else {
					current = current.right
				}
			}
		}
    return this
  }

  contains(value) {
    let current = this
    while (current !== null) {
			if (value < current.value) {
				current = current.left
			} else if (value > current.value) {
				current = current.right
			} else {
				return true
			}
    }
    return false
  }

  remove(value, parent = null) {
		let current = this
		while (current !== null) {
			if (value < current.value) {
				parent = current
				current = current.left
			} else if (value > current.value) {
				parent = current
				current = current.right
			} else {
				if (current.left !== null && current.right !== null) {
					current.value = current.right.getMinValue()
					current.right.remove(current.value, current)
				} else if (parent === null) {
					if (current.left !== null) {
						current.value = current.left.value
						current.right = current.left.right
						current.left = current.left.left
					} else if (current.right !== null) {
						current.value = current.right.value
						current.left = current.right.left
						current.right = current.right.right
					} else {
						current.value = null
					}
				} else if (parent.left === current) {
					parent.left = current.left !== null ? current.left : current.right
				} else if (parent.right === current) {
					parent.right = current.left !== null ? current.left : current.right
				}
				break
			}
		}
    return this
  }

	getMinValue() {
		let current = this
		while (current.left !== null) {
			current = current.left
		}
		return current.value
	}
}

File: binary-search-trees/BST-class.py
class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    # Average: O(log(n)) time | O(1) space
    # Worst: O(n) time | O(1) space
    def insert(self, value):
        current = self
        while True:
            if value < current.value:
                if current.left is None:
                    current.left = BST(value)
                    break
                else:
                    current = current.left
            else:
                if current.right is None:
                    current.right = BST(value)
                    break
                else:
                    current = current.right
        return self

    # Average: O(log(n)) time | O(1) space
    # Worst: O(n) time | O(1) space
    def contains(self, value):
        current = self
        while current is not None:
            if value < current.value:
                current = current.left
            elif value > current.value:
                current = current.right
      else:
                return True
        return False

    # Average: O(log(n)) time | O(1) space
    # Worst: O(n) time | O(1) space
    def remove(self, value, parent = None):
        current = self
        while current is not None:
      if value < current.value:
        parent = current
        current = current.left
      elif value > current.value:
        parent = current
        current = current.right
      else:
        if current.left is not None and current.right is not None:
          current.value = current.right.getMinValue()
          current.right.remove(current.value, current)
        elif parent is None:
          if current.left is not None:
            current.value = current.left.value
            current.right = current.left.right
            current.left = current.left.left
          elif current.right is not None:
            current.value = current.right.value
            current.left = current.right.left
            current.right = current.right.right
          else:
            current.value = None
        elif parent.left == current:
          parent.left = current.left if current.left is not None else current.right
        elif parent.right == current:
          parent.right = current.left if current.left is not None else current.right
        break
        return self

  def getMinValue(self):
    current = self
    while current.left is not None:
      current = current.left
    return current.value
